File: 240206/0206_Mission/KYJmission03.py
import RPi.GPIO as GPIO
import time

buzzer = 18

button_pin1 = 22
button_pin2 = 23
button_pin3 = 24
button_pin4 = 25

led_red = 4
led_green = 5
led_blue = 6

GPIO.setmode(GPIO.BCM) #핀모드 설정
GPIO.setwarnings(False)

GPIO.setup(buzzer, GPIO.OUT)
GPIO.setup(led_red, GPIO.OUT)
GPIO.setup(led_green, GPIO.OUT)
GPIO.setup(led_blue, GPIO.OUT)

GPIO.setup(button_pin1, GPIO.IN, pull_up_down=GPIO.PUD_UP)
GPIO.setup(button_pin2, GPIO.IN, pull_up_down=GPIO.PUD_UP)
GPIO.setup(button_pin3, GPIO.IN, pull_up_down=GPIO.PUD_UP)
GPIO.setup(button_pin4, GPIO.IN, pull_up_down=GPIO.PUD_UP)

GPIO.output(led_red,False)
GPIO.output(led_green,False)
GPIO.output(led_blue,False)

pwm = GPIO.PWM(buzzer, 1.0)

def Music_One():
    pwm.start(10.0)
    scale = [262, 294, 330, 349, 392, 440, 494]
    twinkle = [1,1,5,5,6,6,5,4,4,3,3,2,2,1,5,5,4,4,3,3,2,5,5,4,4,3,3,2,1,1,5,5,6,6,5,4,4,3,3,2,2,1]

    try:
        for i in range(0,42):
            pwm.ChangeFrequency(scale[twinkle[i] - 1])
            GPIO.output(led_red,True)
            if i == 6 or i == 13 or i == 20 or i == 27 or i == 34 or i == 41:
                time.sleep(1)
            else:
                time.sleep(0.5)
        pwm.stop()
        GPIO.output(led_red,False)
    except KeyboardInterrupt:
        pwm.stop()
        GPIO.output(led_red,False)

def Music_Two():
    pwm.start(10.0)
    scale = [523, 587, 659, 698, 784, 880, 988, 1047]
    twinkle = [4,2,2,
               3,1,1,
               0,1,2,3,4,4,4,
               4,2,2,2,3,1,1,
               0,2,4,4,2,2,2,
               1,1,1,1,1,2,3,
               2,2,2,2,2,3,4,
               4,2,2,
               3,1,1,
               0,2,4,4,2,2,2]

    try:
        for i in range(0,54):
            pwm.ChangeFrequency(scale[twinkle[i]])
            GPIO.output(led_green,True)
            if i == 2 or i == 5 or i == 12 or i == 19 or i == 26 or i == 33 or i == 40 or i == 43 or i == 46 or i == 53:
                time.sleep(1)
            else:
                time.sleep(0.5)
        pwm.stop()
        GPIO.output(led_green,False)
    except KeyboardInterrupt:
        pwm.stop()
        GPIO.output(led_green,False)

def Music_Three():
    pwm.start(10.0)
    scale = [523, 587, 659, 698, 784, 880, 988, 1047]
    twinkle = [4,4,5,5,4,4,2,
               4,4,2,2,1,
               4,4,5,5,4,4,2,
               4,2,1,2,0] 

    try:
        for i in range(0,24):
            pwm.ChangeFrequency(scale[twinkle[i]])
            GPIO.output(led_blue,True)
            if i == 6 or i == 18:
                time.sleep(1)
            elif i == 0 or i == 1 or i == 2 or i == 3 or i == 4 or i == 5 or i == 7 or i == 8 or i == 9 or i == 10 or i == 12 or i == 13 or i == 14 or i == 15 or i == 16 or i == 17 or i == 19 or i == 20 or i == 21 or i == 22:
                time.sleep(0.5) 
            elif i == 11 or i == 23:
                time.sleep(1.2)
        pwm.stop()
    except KeyboardInterrupt:
        pwm.stop()
        GPIO.output(led_blue,False)
       

while True:
    if GPIO.input(button_pin1) == GPIO.LOW:
        print("Button 1 pushed!")
        Music_One()
    elif GPIO.input(button_pin2) == GPIO.LOW:
        print("Button 2 pushed!")
        Music_Two()
    elif GPIO.input(button_pin3) == GPIO.LOW:
        print("Button 3 pushed!")
        Music_Three()
    elif GPIO.input(button_pin4) == GPIO.LOW:
        print("Button 4 pushed!")
        GPIO.output(led_red,False)
        GPIO.output(led_green,False)
        GPIO.output(led_blue,False)
        break
GPIO.cleanup()

File: 240206/0206_Mission/KYJmission03.c
#include <signal.h> //Signal 사용 헤더파일
#include <stdio.h>
#include <stdlib.h>
#include <wiringPi.h>
#include <unistd.h>
#include <time.h>

#define SW1 3
#define SW2 4
#define SW3 5
#define SW4 6

#define led_red 7
#define led_green 21
#define led_blue 22

const int Buzzer = 1;

void sig_handler(int signo); // 마지막 종료 함수 ctlr-C 로 종료

void Music_One(void) {
    pwmWrite(Buzzer, 10);
    int scale[] = {523, 587, 659, 698, 784, 880, 988, 1047};
    int twinkle[] = {1,1,5,5,6,6,5,4,4,3,3,2,2,1,5,5,4,4,3,3,2,5,5,4,4,3,3,2,1,1,5,5,6,6,5,4,4,3,3,2,2,1};

    for (int i = 0; i < 42; i++) {
        pwmSetRange(1000000/scale[twinkle[i]]);
        pwmWrite(Buzzer, 1000000/scale[twinkle[i]]/10);
        digitalWrite(led_red, 1);
        if (i == 6 || i == 13 || i == 20 || i == 27 || i == 34 || i == 41) {
            delay(1000);
        }
        else {
            delay(500);
        }
    }
    pwmWrite(Buzzer, 0);
    digitalWrite(led_red, 0);
}

void Music_Two(void) {
    pwmWrite(Buzzer, 10);
    int scale[] = {523, 587, 659, 698, 784, 880, 988, 1047};
    int twinkle[] = {4,2,2,
               3,1,1,
               0,1,2,3,4,4,4,
               4,2,2,2,3,1,1,
               0,2,4,4,2,2,2,
               1,1,1,1,1,2,3,
               2,2,2,2,2,3,4,
               4,2,2,
               3,1,1,
               0,2,4,4,2,2,2};

    for (int i = 0; i < 54; i++) {
        pwmSetRange(1000000/scale[twinkle[i]]);
        pwmWrite(Buzzer, 1000000/scale[twinkle[i]]/10);
        digitalWrite(led_green, 1);
        if (i == 2 || i == 5 || i == 12 || i == 19 || i == 26 || i == 33 || i == 40 || i == 43 || i == 46 || i == 53) {
            delay(1000);
        }
        else {
            delay(500);
        }
    }
    pwmWrite(Buzzer, 0);
    digitalWrite(led_green, 0);
}

void Music_Three(void) {
    pwmWrite(Buzzer, 10);
    int scale[] = {523, 587, 659, 698, 784, 880, 988, 1047};
    int twinkle[] = {4,4,5,5,4,4,2,
               4,4,2,2,1,
               4,4,5,5,4,4,2,
               4,2,1,2,0};

    for (int i = 0; i < 24; i++) {
        pwmSetRange(1000000/scale[twinkle[i]]);
        pwmWrite(Buzzer, 1000000/scale[twinkle[i]]/10);
        digitalWrite(led_blue, 1);
        if (i == 6 || i == 18) {
            delay(1000);
        }
        else if (i == 0 || i == 1 || i == 2 || i == 3 || i == 4 || i == 5 || i == 7 || i == 8 || i == 9 || i == 10 || i == 12 || i == 13 || i == 14 || i == 15 || i == 16 || i == 17 || i == 19 || i == 20 || i == 21 || i == 22) {
            delay(500);
        }
        else if (i == 11 || i == 23) {
            delay(1200);
        }
    }
    pwmWrite(Buzzer, 0);
    digitalWrite(led_blue, 0);
}

// 첫번째 스위치 인터럽트 처리 함수
void sw1_isr(void) {
  printf("Music One!\n"); // 스위치 눌림을 출력
  Music_One();
}

// 두번째 스위치 인터럽트 처리 함수
void sw2_isr(void) {
  printf("Music Two!\n"); // 스위치 눌림을 출력
  Music_Two();
}

// 세번째 스위치 인터럽트 처리 함수
void sw3_isr(void) {
  printf("Music Three!\n"); // 스위치 눌림을 출력
  Music_Three();
}

// 네번째 스위치 인터럽트 처리 함수
void sw4_isr(void) {
  printf("SW4 pressed\n"); // 스위치 눌림을 출력
  printf("Program terminated\n"); // 프로그램 종료를 출력
  exit(0); // 프로그램 종료
}

int main(void) {
    signal(SIGINT, (void *)sig_handler); //시그널 핸들러 함수

    if(wiringPiSetup () == -1)
        return 1;
    pinMode(Buzzer, PWM_OUTPUT);
    pinMode(led_red, OUTPUT);
    pinMode(led_green, OUTPUT);
    pinMode(led_blue, OUTPUT);
    pinMode(SW1, INPUT);
    pinMode(SW2, INPUT);
    pinMode(SW3, INPUT);
    pinMode(SW4, INPUT);
    digitalWrite(led_red, 0);
    digitalWrite(led_green, 0);
    digitalWrite(led_blue, 0);
    pwmSetClock(19);
    pwmSetMode(PWM_MODE_MS);
    wiringPiISR(SW1, INT_EDGE_FALLING, sw1_isr); // 첫번째 스위치가 눌리면 sw1_isr 함수 호출
    wiringPiISR(SW2, INT_EDGE_FALLING, sw2_isr); // 두번째 스위치가 눌리면 sw2_isr 함수 호출
    wiringPiISR(SW3, INT_EDGE_FALLING, sw3_isr); // 세번째 스위치가 눌리면 sw3_isr 함수 호출
    wiringPiISR(SW4, INT_EDGE_FALLING, sw4_isr); // 네번째 스위치가 눌리면 sw4_isr 함수 호출
    while (1) {
    }
    return 0;
}

void sig_handler(int signo)  //마지막 처리를 위한 signal 핸들러 함수 모두 OFF
{
    printf("process stop\n");
    pwmWrite(Buzzer, 0); // Off
    digitalWrite(led_red, 0);
    digitalWrite(led_green, 0);
    digitalWrite(led_blue, 0);
    
    exit(0);
}